File: ios/App/Snappet/Features/Budget/BudgetModels.swift
import Foundation
import SwiftData

/// A spending category with a monthly limit (e.g. "Groceries", $400). Keyed by a stable
/// `id` (UUID) so transactions reference it independently of SwiftData object identity.
@Model
final class BudgetCategory {
    /// Stable identity used to key `BudgetTransaction` records.
    var id: UUID
    var name: String
    /// The budgeted spend ceiling for one calendar month.
    var monthlyLimit: Double
    var createdAt: Date

    init(id: UUID = UUID(), name: String, monthlyLimit: Double, createdAt: Date = .now) {
        self.id = id
        self.name = name
        self.monthlyLimit = monthlyLimit
        self.createdAt = createdAt
    }
}

/// A single spend logged against a category. `categoryID` matches `BudgetCategory.id`
/// so categories can be deleted/recreated without dangling object references.
@Model
final class BudgetTransaction {
    /// Matches `BudgetCategory.id`.
    var categoryID: UUID
    var amount: Double
    var note: String
    var date: Date

    init(categoryID: UUID, amount: Double, note: String = "", date: Date = .now) {
        self.categoryID = categoryID
        self.amount = amount
        self.note = note
        self.date = date
    }
}

File: android/app/src/main/java/com/snappet/mobile/feature/budget/BudgetModels.kt
package com.snappet.mobile.feature.budget

import androidx.room.Entity
import androidx.room.PrimaryKey

/**
 * A spending category with a monthly limit (e.g. "Groceries", 400/mo). Keyed by a stable
 * [categoryId] (UUID string) so [BudgetTransaction] records reference it independently of the
 * Room row id (categories can be deleted/recreated without dangling references). Mirrors the iOS
 * `BudgetCategory` `@Model`.
 */
@Entity(tableName = "budget_categories")
data class BudgetCategory(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    /** Stable identity used to key [BudgetTransaction] records. */
    val categoryId: String = java.util.UUID.randomUUID().toString(),
    val name: String,
    /** The budgeted spend ceiling for one calendar month. */
    val monthlyLimit: Double,
    /** When the category was created (epoch millis). */
    val createdAt: Long,
)

/**
 * A single spend logged against a category. [categoryId] matches [BudgetCategory.categoryId] so
 * categories can be deleted/recreated without dangling object references. Mirrors the iOS
 * `BudgetTransaction` `@Model`.
 */
@Entity(tableName = "budget_transactions")
data class BudgetTransaction(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    /** Matches [BudgetCategory.categoryId]. */
    val categoryId: String,
    val amount: Double,
    val note: String = "",
    /** When the spend occurred (epoch millis). */
    val date: Long,
)
